FirebaseAI/Sources/Types/Internal/Live/AsyncWebSocket.swift

```diff
@@ -0,0 +1,118 @@
+// Copyright 2025 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+import Foundation
+
+final class AsyncWebSocket: NSObject, @unchecked Sendable, URLSessionWebSocketDelegate {
+  private let webSocketTask: URLSessionWebSocketTask
+  private let stream: AsyncThrowingStream<URLSessionWebSocketTask.Message, Error>
+  private let continuation: AsyncThrowingStream<URLSessionWebSocketTask.Message, Error>.Continuation
+  private var continuationFinished = false
+  private let continuationLock = NSLock()
+
+  private var _closeError: WebSocketClosedError? = nil
+  private let closeErrorLock = NSLock()
+  private(set) var closeError: WebSocketClosedError? {
+    get { closeErrorLock.withLock { _closeError } }
+    set { closeErrorLock.withLock { _closeError = newValue } }
+  }
```

Comment:
There is a type in FirebaseCoreInternal for doing this kind of locking.

```
FirebaseCore/Internal/Sources/Utilities/UnfairLock.swift:public final class UnfairLock<Value>: @unchecked Sendable {
```
Should be available with a  `private import FirebaseCoreInternal` import.

Reply:
While slightly different, this is a good api. I've adapted it:)